adhocteam/Head-Start-TTADP
# Add collaborators

**Description of change**

Includes a few additional improvements, got bigger because I refactored some tests to hopefully make them easier to deal with in the future.

 * Collaborators join table added.
 * Collaborators hooked up on the frontend. Available collaborators depends on the report's region (hardcoded to 1 for now)
 * Activity report checks permissions
 * Added the first "policies" which can be used for authorization. Should be easier to test separated from controllers
 * Route tests that hit the DB moved to the services folder, only deal with the DB interactions
 * Route tests mock out DB and authorization calls focusing on constructing correct responses based on (mocked) calls to the DB/policies
 * Multiselect sets null values to an empty array, allowing collaborators/recipients to be "zeroed" in the report
 * Approvers are scoped to region
 * The activity report now makes all API calls up front in parallel. Should keep other components simpler and cuts down on calls to the API (they can be made once instead on every form page load)
 * Added Krys and I to seeders as Admins so we should be admins in sandbox/dev until HSES resets again. (can add others to the seeders but we were the only two who currently have accounts on sandbox, so the only two I know HSES ids)
 * Updated swagger/API docs

**How to test**

1) Browse to [sandbox](https://tta-smarthub-sandbox.app.cloud.gov/activity-reports/new)
2) Fill out a report, add Krys or Josh as a collaborator (If you aren't on the list I can add you, just ping me)
3) Verify collaborators persist across reloads
4) Send the ID of the report to someone who isn't a collaborator and verify they can't access the report

**Note**
I still need to update API docs/DB docs

**Issue(s)**
* https://github.com/HHS/Head-Start-TTADP/issues/109

**Checklist**

    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified
